src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/ace/AceEditorIdleCommands.java

```diff
@@ -154,6 +168,250 @@ private void onPreviewImage(AceEditor editor,
       panel.show();
    }
    
+   private void onPreviewImageLineWidget(final AceEditor editor,
+                                         final String href,
+                                         final Position position,
+                                         final Range tokenRange)
+   {
+      // if we already have a line widget for this row, bail
+      LineWidget lineWidget = editor.getLineWidgetForRow(position.getRow());
+      if (lineWidget != null)
+         return;
+      
+      // shared mutable state that we hide in this closure
+      final Mutable<PinnedLineWidget>  plw = new Mutable<PinnedLineWidget>();
+      final Mutable<ChunkOutputWidget> cow = new Mutable<ChunkOutputWidget>();
+      final Mutable<HandlerRegistration> docChangedHandler = new Mutable<HandlerRegistration>(); 
+      final Mutable<HandlerRegistration> renderHandler = new Mutable<HandlerRegistration>();
+      
+      // command that ensures state is cleaned up when widget hidden
+      final Command onDetach = new Command()
+      {
+         private void detach()
+         {
+            // detach chunk output widget
+            cow.set(null);
+
+            // detach pinned line widget
+            if (plw.get() != null)
+               plw.get().detach();
+            plw.set(null);
+
+            // detach render handler
+            if (renderHandler.get() != null)
+               renderHandler.get().removeHandler();
+            renderHandler.set(null);
+
+            // detach doc changed handler
+            if (docChangedHandler.get() != null)
+               docChangedHandler.get().removeHandler();
+            docChangedHandler.set(null);
+         }
+         
+         @Override
+         public void execute()
+         {
+            // if the associated chunk output widget has been cleaned up,
+            // make a last-ditch detach effort anyhow
+            ChunkOutputWidget widget = cow.get();
+            if (widget == null)
+            {
+               detach();
+               return;
+            }
+            
+            // fade out and then detach
+            FadeOutAnimation anim = new FadeOutAnimation(widget, new Command()
+            {
+               @Override
+               public void execute()
+               {
+                  detach();
+               }
+            });
+            
+            anim.run(400);
+         }
+      };
+      
+      // construct placeholder for image
+      final SimplePanel container = new SimplePanel();
+      final Label noImageLabel = new Label("(No image at path " + href + ")");
```

Comment:
Do we need to escape the href here? (Can't recall offhand if Label uses innerText)